### PR TITLE
Refine logging output

### PR DESCRIPTION
**Pull request by atgeirr**
This adds the possibility to configure how any back-end should decorate messages. Decoration here means adding a prefix and/or changing color depending on message type. The default is to have no decoration.

Actual decoration is handled by the concrete log backend calling the decorateMessage() protected method from the LogBackend superclass. So far only the StreamLog does this (it is the only log actually doing any output, all others inherit it).

Behavior should be preserved for now, with the exception of disappearing prefixes (like "warning: ") before log messages (unless client code configures the backend to use them).

I did some cleanup and docs for the LogBackend class, in the first commit, which can make reading the diff for that class a little unclear.

### CODE REVIEW

**Comment by bska** [comment]
I have a couple of reservations about the colour aspect of this.  First, the ASCII escape sequences are as non-portable as you can possibly get.  This must never be enabled by default.  Second, unless I'm completely misreading the patch, the test doesn't actually test anything--save for the order of function calls.  You appear to be using the same function calls to construct the `expected` string as the string being expected.  This approach cannot uncover any issues in the individual functions.

**Comment by atgeirr** [comment]
> First, the ASCII escape sequences are as non-portable as you can possibly get. This must never be enabled by default

I think that we should deal with this on a program-by-program basis, but I'd like to consider having it on by default for Flow for example. We could add a few more safeguards though, such as testing that we are really on a terminal. I believe that all relevant terminals in use today would support this, however. 

> This approach cannot uncover any issues in the individual functions

Right, that would call for a separate test for those functions themselves, which does not exist at this point. I'll add it.

**Comment by bska** [comment]
>> First, the ASCII escape sequences are as non-portable as you can possibly get. This must never be enabled by default
> 
> I think that we should deal with this on a program-by-program basis, but I'd like to consider having it on by default for Flow for example. We could add a few more safeguards though, such as testing that we are really on a terminal. I believe that all relevant terminals in use today would support this, however.

To be honest, I'd rather not have `Opm::Log` deal with colours and terminals at all.  This is best relegated to a Python-based postprocessor for the log.

**Comment by atgeirr** [comment]
> This is best relegated to a Python-based postprocessor for the log.

That requires us in any case to decorate messages in some way. I imagine that for the file lof prefix decoration will be preferred for this very reason.

**Comment by atgeirr** [comment]
Added tests for helper functions.

**Comment by atgeirr** [comment]
jenkins build this with downstreams please

**Comment by bska** [comment]
> I imagine that for the file lof prefix decoration will be preferred for this very reason.

Sorry, I don't understand what you mean.  Would care to rephrase the above?  As for log entry decoration, That Other Simulator already does this using strings of the form
```
@<TYPE>
```
e.g., `@WARNING`.  This is easy to parse and deal with in a post-processor.

**Comment by atgeirr** [comment]
> Sorry, I don't understand what you mean

"lof" should be "log"...

> e.g., @WARNING. This is easy to parse and deal with in a post-processor.

That is a nice and simple scheme, and we should adopt it.

**Comment by qilicun** [comment]
It Works! But maybe just set prefix with the different color should be sufficient?
Following is the test from ``relperm_diagnostics``

screenshot from 2016-05-10 09 34 32

**Comment by atgeirr** [comment]
> But maybe just set prefix with the different color should be sufficient?

I am open to all kinds of refinements on this, my main point was to treat decoration of messages in a unified way.

I do suggest that this can be merged as-is though to get the basic framework in (with only minimal behavior changes), and then it can be refined later.

**Comment by qilicun** [comment]
>I do suggest that this can be merged as-is though to get the basic framework in (with only minimal behavior changes), and then it can be refined later.

Sure, and it works as expected, we can improve it later.

**Comment by atgeirr** [comment]
@joakim-hove please review

**Comment by joakim-hove** [comment]
Will review today - but quite tight schedule, so not immediately.

File: opm/common/OpmLog/LogUtil.cpp [hunk 1]
```diff
@@ -72,5 +72,29 @@ namespace Log {
...
         return prefix + ": " + message;
     }
+
+
+    std::string colorCodeMessage(int64_t messageType, const std::string& message) {
+        std::string colorcode;
+        switch (messageType) {
+        case MessageType::Debug:
+        case MessageType::Info:
+            colorcode = "\033[39m";
```
Comment (joakim-hove) [inline opm/common/OpmLog/LogUtil.cpp]: Hehe  - I recognize a magic string when I see one! Please some descriptive names:
```C++
const std::string magenta = "\033[39m";
const std::string color_off = "\033[0m";
```

File: opm/common/OpmLog/Logger.cpp [hunk 2]
```diff
@@ -63,6 +63,11 @@ namespace Opm {
             return true;
     }
 
+    void Logger::clearBackends() {
```
Comment (joakim-hove) [inline opm/common/OpmLog/Logger.cpp]: Can I suggest `drop / delete / remove / ...` the name `clear()` seems to indicate that only the internal state of the backends is cleared/reset.

File: opm/common/OpmLog/LogBackend.cpp [hunk 3]
```diff
@@ -27,18 +27,37 @@ namespace Opm {
...
+    {
         return m_mask;
     }
 
+    bool LogBackend::includeMessage(int64_t messageFlag)
+    {
+        return ((messageFlag & m_mask) == messageFlag) && (messageFlag > 0);
+    }
+
+    std::string LogBackend::decorateMessage(int64_t messageFlag, const std::string& message)
```
Comment (joakim-hove) [inline opm/common/OpmLog/LogBackend.cpp]: If the decoration process is based on a private method like:
```C++
std::string decorate_impl( const std::string& msg, const std::string& prefix , const std::string& suffix = "") {
     return prefix + msg + suffix;
}  
```
then both color coding and string based prefix / suffix can be implemented with the same code.

**Comment by jokva** [comment]
Why didn't we end up going for `@WARNING` style output and let post processor or filters deal with colours?

File: opm/common/OpmLog/Logger.cpp (hunk 2 above)
Comment (atgeirr) [inline opm/common/OpmLog/Logger.cpp]: What about `removeAllBackends()`?

Comment (joakim-hove) [inline opm/common/OpmLog/Logger.cpp]: > What about removeAllBackends()?

Sure!

**Comment by atgeirr** [comment]
> Why didn't we end up going for @WARNING style output and let post processor or filters deal with colours?

My primary intention was just to add a framework, I'll happily support the @WARNING style as a separate PR. I do think colour is very useful in terminal output so I'd like that to be possible. Also: while we need to support post processors, *requiring* post processors/filters to get the most useful output is a no-go I think.

About the @WARNING scheme: from my (limited) exposure to PRT files it seems that it is a little more involved than just a simple prefix, it looks more like this: `@--WARNING\n@ <12 spaces>`, followed by word-wrapping and flowing the message text to start each line with the same indent.

File: tests/test_OpmLog.cpp [hunk 4]
```diff
@@ -249,3 +232,71 @@ BOOST_AUTO_TEST_CASE(TestOpmLog) {
...
+
+    {
+        std::shared_ptr<CounterLog> counter = std::make_shared<CounterLog>();
+        std::shared_ptr<StreamLog> streamLog = std::make_shared<StreamLog>(log_stream, Log::DefaultMessageTypes);
+        BOOST_CHECK_EQUAL( false , OpmLog::hasBackend("NO"));
+        OpmLog::addBackend("COUNTER" , counter);
+        OpmLog::addBackend("STREAM" , streamLog);
+        BOOST_CHECK_EQUAL( true , OpmLog::hasBackend("COUNTER"));
+        BOOST_CHECK_EQUAL( true , OpmLog::hasBackend("STREAM"));
+
```
Comment (joakim-hove) [inline tests/test_OpmLog.cpp]: `BOOST_CHECK_EQUAL( textColor ,` :red_circle: `)`

 :smile:

Comment (atgeirr) [inline tests/test_OpmLog.cpp]: Sorry did not catch the joke (if it was one)...?

File: opm/common/OpmLog/LogBackend.cpp [hunk 5]
```diff
@@ -27,18 +27,37 @@ namespace Opm {
...
+    {
+    }
 
-    bool LogBackend::includeMessage(int64_t messageFlag) {
-        if (((messageFlag & m_mask) == messageFlag) &&
-            (messageFlag > 0))
-            return true;
-        else
-            return false;
+    void LogBackend::configureDecoration(bool use_prefix, bool use_color_coding)
```
Comment (joakim-hove) [inline opm/common/OpmLog/LogBackend.cpp]: With the risk of going overboard; we could implement a `std::unorderd_map<int64_t , std::string>`with per message type prefix (and color) configuration.

```C++
void LogBackend::configurePrefix( int64_t messageType , const std::string& prefix) {

}

void LogBackend::configureColor(int64_t messageType, const ColorType& color)  {

}
```

File: tests/test_OpmLog.cpp (hunk 4 above)
Comment (joakim-hove) [inline tests/test_OpmLog.cpp]: Well - sure it was a joke - looked kind of cute in my browser!

File: opm/common/OpmLog/LogBackend.cpp (hunk 5 above)
Comment (atgeirr) [inline opm/common/OpmLog/LogBackend.cpp]: I like the idea, but I think it is a little too much for the poor LogBackend, when we need this kind of detail we should perhaps make a separate MessageDecorator or MessageFormatter class?

File: tests/test_OpmLog.cpp (hunk 4 above)
Comment (atgeirr) [inline tests/test_OpmLog.cpp]: It did... :-)

**Comment by joakim-hove** [comment]
OK; to summarize:

1. The non decoration based changes are fine - happy to merge.
2. I just can not merge code with the magic string `"033m[81]"`!
3. If this really should be a *framework* - I would suggest taking it just a bit further - like: https://github.com/OPM/opm-common/pull/115#discussion_r62979314 - that way the scheme `@WARNING` can be configured simply.

**Comment by atgeirr** [comment]
I have now adressed the identified issues (I think). I added an abstract class for message formatters, so that those of you with detailed knowledge of how it should work can create a good PRTMessageFormatter class for example (I did not at this point).

File: opm/common/OpmLog/LogBackend.cpp [hunk 6]
```diff
@@ -27,18 +27,33 @@ namespace Opm {
...
 
-    bool LogBackend::includeMessage(int64_t messageFlag) {
-        if (((messageFlag & m_mask) == messageFlag) &&
-            (messageFlag > 0))
-            return true;
-        else
-            return false;
+    void LogBackend::configureDecoration(std::shared_ptr<MessageFormatterInterface> formatter)
+    {
+        formatter_ = formatter;
```
Comment (joakim-hove) [inline opm/common/OpmLog/LogBackend.cpp]: Ahhh - we are going to fight a long time on those trailing underscores!

**Comment by joakim-hove** [comment]
OK; this looks good - now only two comments left:

1. Would it not be be natural to have the color coding implemented in the `MessageFormatter.cpp` instead of `LogUtil.cpp`?
2. Every trailing `_` on member names will cost you a `{ ... }` pair :grinning:

File: opm/common/OpmLog/LogBackend.cpp (hunk 6 above)
Comment (atgeirr) [inline opm/common/OpmLog/LogBackend.cpp]: Why?

Comment (jokva) [inline opm/common/OpmLog/LogBackend.cpp]: Because `this->` is clearer, compiler checked and doesn't encourage redundancy in parameter naming.

**Comment by atgeirr** [comment]
> Would it not be be natural to have the color coding implemented in the MessageFormatter.cpp instead of LogUtil.cpp?

Maybe, I just took the route of minimal change. The prefix and color routines could be used by other parts of the code though, so I'm not sure how much is to gain.

File: opm/common/OpmLog/LogBackend.cpp (hunk 6 above)
Comment (atgeirr) [inline opm/common/OpmLog/LogBackend.cpp]: But `this->` is not enforced in any way. Also I don't understand
> doesn't encourage redundancy in parameter naming

This has been a very long-standing standard for many codes, not just OPM, so if you want me to adopt a new point of view, point me to the links I should read :-)

**Comment by joakim-hove** [comment]
> Maybe, I just took the route of minimal change.

Fair enough - merging.

File: opm/common/OpmLog/LogBackend.cpp (hunk 6 above)
Comment (joakim-hove) [inline opm/common/OpmLog/LogBackend.cpp]: > so if you want me to adopt a new point of view, point me to the links I should read :-)

Well - just like unnecessary braces it just looks ugly!  Seriously I have actually though a bit about your comments about the beauty of heterogeneity also in code - I kind of liked that view.

Comment (jokva) [inline opm/common/OpmLog/LogBackend.cpp]: Neither is trailing underscore, but using `this->` on a non-member is an error, but naming a non-member with trailing underscore is not. Also **removes** the risk of accessing a shadowing variable.

> This has been a very long-standing standard for many codes, not just OPM, so if you want me to adopt a new point of view, point me to the links I should read :-)

Yes, and Hungarian notation was popular for a while. This is not an argument on trailing underscore's merit, only its adoption.

See, what happens with _ is that there are plenty of functions that go something like

```void foo::bar( int var ) {
    var_ += var;
}
```
See, the name `var` shows up twice with a fairly easy-to-miss modification. Way too often this happens too:
`void Foo::bar( Baz baz ) { baz_ = baz }` 

Which is somewhat leaky, but more importantly baz_ and baz aren't all that different. Also, why does the name `baz` in the argument *have* to match `Foo`'s member's name?

I'd typically opt for a different approach:
`void Foo::bar( Baz x ) { this->baz = x }`

with `x` named something applicable to the context or *intent* of change. Also, it's **obvious** how this modifies the objects internal state, and what is defined and control by the object. baz_ doesn't convey that as well.

(also, I have syntax highlighting on reserved keywords, and `this` is reserved).

This could go on forever with bigger and more realistic examples, but its the gist of it.

Comment (atgeirr) [inline opm/common/OpmLog/LogBackend.cpp]: Don't get me wrong: there's not much wrong about `this->` (apart from being a bit wordy in contexts where we invoke multiple members), and I don't think your arguments are without merit. You missed my point however: nothing enforces using `this->`, if I write `data = 7` that is perfectly valid. Do you have external links providing further arguments? (That's what I asked for...)

In any case, I'll enforce the existing rule where I am maintainer, elsewhere other rules will be followed (such as `m_` in opm-parser I guess), we'll just have to accept this and live with it. Keeping things consistent within modules is good, and within single files mandatory. As long as everybody's able to adapt as needed I think there is no big problem.

Comment (jokva) [inline opm/common/OpmLog/LogBackend.cpp]: I never claimed `this->` was enforced; I claimed using `this->` enforces the target being a member.

**Comment by joakim-hove** [comment]
>In any case, I'll enforce the existing rule where I am maintainer, elsewhere other rules will be followed (such as m_ in opm-parser I guess), we'll just have to accept this and live with it. Keeping things consistent within modules is good, and within single files mandatory. 


Well - in this case I would say you have failed to follow your own principles; you have forced in the use of `formatter_` in a file where another convention was already established; but I guess you are willing to pay in `{ ... }` pairs? :grinning:

**Comment by atgeirr** [comment]
> Well - in this case I would say you have failed to follow your own principles; you have forced in the use of formatter_ in a file where another convention was already established

You are absolutely right, I did not notice. If this had been brought to attention before the merge I would have changed it. Sorry about that!

**Comment by bska** [comment]
I'll try my best to stay out of this discussion, but I'd just like to present the following example for why I **always** put braces on my blocks
```C++
if (cond)
    if (other)
        doA();
else
    doB();
```
GCC 6 warns about this, but unless you're using that toolchain it's just easier to always put braces on the blocks to enforce a consistent reading.

**Comment by pgdr** [comment]
@bska I don't think that's what's being discussed here.  I think we can all agree that heavy nesting should go hand-in-hand with braces.  I think the real question is, _what do braces improve over the following code_?

```C++
if (x < 0)
    continue;
```

**Comment by atgeirr** [comment]
For me this comes down to bitter experience, a case I saw not too long ago was:
```
GL_RESULT res = someOpenGlCall();
```
got turned into
```
GL_RESULT res = someOpenGlCall(); SOME_CHECKING_MACRO(res);
```
(typically this pattern is followed over many lines), then this was made conditional:
```
if (should_call)
    GL_RESULT res = someOpenGlCall(); SOME_CHECKING_MACRO(res);
```

I can find your example a bit better looking without braces, but usually things are a bit more convoluted, and a simple rule (always use braces) prevents errors.

**Comment by joakim-hove** [comment]
This discussion needs more: :beer: :beers: :beer: 

But seriously forcing an *always use braces* rule based on an example like

```C++
if (should_call)
    GL_RESULT res = someOpenGlCall(); SOME_CHECKING_MACRO(res);
```

is in my opinion a bit like always wearing a helmet when you drive a car - just because you have seen Formula 1 on TV. 

We all have massive egos and sense of pride in creating readable and maintainable code (I *really* think we agree on the goals here) - and then we just have to accept that we disagree over some details of the - way; I refuse to wear the helmet when driving my 2008 Volvo model with too small engine - but sure I will not deny anyone else doing it (including my passengers!)

**Comment by bska** [comment]
> I think the real question is, what do braces improve over the following code?
```C++
if (x < 0)
    continue;
```

Probably nothing in isolation.  It's just that code changes during maintenance and I don't trust myself to recognise when I have to add braces.  Having perfect test coverage obviously helps.

That said, putting braces in in the first place and then not thinking about it anymore has brought me to a place where a "naked" block just looks fundamentally "unsafe" to my eyes.  There are enough rules to remember in the language so I don't want to spend energy on a case that's entirely preventable by a simple convention.

> I *really* think we agree on the goals [of creating readable and maintainable code]

Agreed.  We may just want to go the whole way of creating an "OPM" formatting style to be enforced by clang-format...
